### PR TITLE
Adding tests for Splunk logging driver

### PR DESCRIPTION
**Pull request by nwang92**
**\- What I did**
Added functional tests for the Splunk logging driver for Docker (https://docs.docker.com/engine/admin/logging/splunk/)

**\- How I did it**
Modified the Makefile to include a test-splunk-logger script. Created test-splunk-logger, which builds a Splunk image (from Dockerfile) with some settings baked in for the tests to execute properly. These settings + Dockerfiles + test script are in a new folder called "loggers". 

**\- How to verify it**
Run "make test" or "make test-splunk-logger".

Signed-off-by: Nelson Wang nwang@splunk.com

### CODE REVIEW

**Comment by icecrime** [comment]
Could the CI failures in the experimental build be related?

**Comment by nwang92** [comment]
Yeah, it looks like 2 of my tests failed on the experimental project due to a connection refused error on the port opened by my Splunk container. I'll make some changes to see if I can make this more robust.

**Comment by nwang92** [comment]
Looks like the CI tests are passing now.

As an aside, there were some previous tests that failed with "No space left on device" errors. It seems to happen infrequently and at different points (during unit tests, my splunk-logger tests, it varies.). I assume that might be a Jenkins related issue, haven't seen it when I run tests locally.

**Comment by thaJeztah** [comment]
Thanks @nwang92, yes it happens on occasion (no space left)

**Comment by LK4D4** [comment]
I'd really prefer if splunk run those tests in their infrastructure, our tests already ridiculuosly slow :/

**Comment by thaJeztah** [comment]
Agreed; would it be good to have the test in this repo, and don't run them by default though?

**Comment by LK4D4** [comment]
I dunno :) even docker-py tests in different repo.

**Comment by nwang92** [comment]
Understandable. I think moving these tests away from being defaults would be ok. The reason for adding it to this repo in particular is to catch any regression due to Docker changes. Also, if someone wanted to add to the logging driver, they would most likely fork it from this repo so it'd be nice to have some tests built-in.

I'll keep "make test-splunk-logger" and remove this test from the "make test" command, if we're all ok with that?

**Comment by thaJeztah** [comment]
We discussed this, and we're okay if this is included in the contrib/ directory; could you move it there?

**Comment by nwang92** [comment]
Done, moved everything to contrib/ and also got removed of a lingering call to these tests in make.sh.

File: Makefile [hunk 1]
```diff
@@ -109,6 +109,9 @@ test-docker-py: build ## run the docker-py tests
 test-integration-cli: build ## run the integration tests
 	$(DOCKER_RUN_DOCKER) hack/make.sh dynbinary test-integration-cli
 
+test-splunk-logger: build ## run the logger tests
```
Comment (crosbymichael) [inline Makefile]: I would rather rename this to something more generic that allows for the testing of all loggers in the future.  

Maybe something like `test-log-drivers`? maybe you have a better idea of a generic name, i'm not totally sold on that one

Comment (tianon) [inline Makefile]: Yeah, I agree -- it's a little strange to have the log driver in contrib, but still have bits in `Makefile` and `hack/make` supporting it.  Is there a way we could instead have a script in `contrib` instead which bootstraps the driver into the codebase?

Comment (nwang92) [inline Makefile]: I think all the log driver code is already in the codebase, the additions for this PR are simply test files and scripts to enable developers to run some tests against the functionality of the logging driver if they want to make any changes to it.

+1 on generalizing the function call, I can rename it.

**Comment by vdemeester** [comment]
hum @nwang92, having trouble running the tests :

```bash
λ make test-log-drivers                                                                                                
docker build  -t "docker-dev:pr-23144" -f "Dockerfile" .
# […]
Successfully installed docker-py splunk-sdk ipaddress requests websocket-client backports.ssl-match-hostname
Cleaning up...
Sending build context to Docker daemon 27.65 kB
Step 1 : FROM outcoldman/splunk:6.4.1
Pulling repository docker.io/outcoldman/splunk
Error: image outcoldman/splunk:6.4.1 not found
---> Making bundle: .integration-daemon-stop (in bundles/1.12.0-dev/test-log-drivers)
+++++ cat bundles/1.12.0-dev/test-log-drivers/docker.pid
++++ kill 4969
make: *** [Makefile:113: test-log-drivers] Error 1
```

**Comment by outcoldman** [comment]
The image this test is based on is not available anymore in Docker Hub. Let's mark this PR as blocked (if this is possible) till the time when Splunk will publish official image on Docker Store

see https://github.com/outcoldman/docker-splunk/issues/23

/cc @mchene

**Comment by thaJeztah** [comment]
>  Let's mark this PR as blocked (if this is possible) 

We don't currently have a process to mark a PR as blocked; overall we just close the PR until it's unblocked; @nwang92 do you think we should close the PR for now? It's best if you close it yourself, because if a maintainer closes it, you're not able to reopen it yourself when it's ready to be reopened

**Comment by nwang92** [comment]
Sure thing, I'll close this for now and reopen when it's ready again.

**Comment by thaJeztah** [comment]
Thanks @nwang92

**Comment by outcoldman** [comment]
@nwang92 images are public now under https://hub.docker.com/u/splunk/ you can use enterprisetrial image

**Comment by mlaventure** [comment]
The base image is not accessible from hub.

```
Step 1/5 : FROM splunk/enterprisetrial:latest
Pulling repository docker.io/splunk/enterprisetrial
Error: image splunk/enterprisetrial:latest not found
```

ping @nwang92 @outcoldman

**Comment by thaJeztah** [comment]
Could it be it was renamed again? https://hub.docker.com/r/splunk/splunk/

**Comment by mlaventure** [comment]
Also means this would need its own CI job to ensure it doesn't break. Right now without testing it manually, they're no way to check that it passes.

**Comment by mchene** [comment]
@thaJeztah Yes, the hub repo for the free Splunk Enterprise was renamed to the link you provided.  All tests should run against the latest release of this repo.

**Comment by thaJeztah** [comment]
Thanks @mchene 

@nwang92 can you update the PR?

**Comment by nwang92** [comment]
Updated, try again with "make test-log-drivers"

**Comment by mlaventure** [comment]
@nwang92 this is still failing for me: https://gist.github.com/mlaventure/eae822cb9816c824c7139101334f5fae

**Comment by nwang92** [comment]
Sorry for the delayed response, seems like the Splunk container didn't spin up because of: "docker-proxy" executable file not found in $PATH. I'm looking into this now.

**Comment by nwang92** [comment]
@mlaventure Having trouble reproducing your exact commands. I tried "hack/make.sh test-log-drivers" but I'm getting an error (error still persists even if I run "hack/make.sh binary-client" prior to the test command): https://gist.github.com/nwang92/c7021ef614074b0d8102ecfdebafd0cf

I am able to run tests if I add in a binary command like "hack/make.sh binary test-log-drivers". See logs: https://gist.github.com/nwang92/f97ee4f0fbe10c64284cb0fb0e99b914. Can you try that and let me know if you're seeing passing tests?

**Comment by mlaventure** [comment]
@nwang92, it still doesn't work for me (I rebased on master first, but it didn't cause any conflicts): `repository outcoldman/splunk not found: does not exist or no read access`

The full log is below.

```
root@39532f490be8:/go/src/github.com/docker/docker# ./hack/make.sh dynbinary test-log-drivers

bundles/1.14.0-dev already exists. Removing.

---> Making bundle: dynbinary (in bundles/1.14.0-dev/dynbinary)
Building: bundles/1.14.0-dev/dynbinary-client/docker-1.14.0-dev
Created binary: bundles/1.14.0-dev/dynbinary-client/docker-1.14.0-dev
Building: bundles/1.14.0-dev/dynbinary-daemon/dockerd-1.14.0-dev
Created binary: bundles/1.14.0-dev/dynbinary-daemon/dockerd-1.14.0-dev

---> Making bundle: test-log-drivers (in bundles/1.14.0-dev/test-log-drivers)
---> Making bundle: .integration-daemon-start (in bundles/1.14.0-dev/test-log-drivers)
+++ /etc/init.d/apparmor start
Starting AppArmor profiles:Warning from stdin (line 1): /sbin/apparmor_parser: cannot use or update cache, disable, or force-complain via stdin
.
INFO: Waiting for daemon to start...
+++ exec dockerd --debug --host unix:///go/src/github.com/docker/docker/bundles/1.14.0-dev/test-log-drivers/docker.sock --storage-driver vfs --pidfile bundles/1.14.0-dev/test-log-drivers/docker.pid --userland-proxy=true
.
Requirement already satisfied (use --upgrade to upgrade): pytest in /usr/local/lib/python2.7/dist-packages (from -r contrib/loggers/splunk/requirements.txt (line 1))
Downloading/unpacking docker-py (from -r contrib/loggers/splunk/requirements.txt (line 2))
Downloading/unpacking splunk-sdk (from -r contrib/loggers/splunk/requirements.txt (line 3))
  Running setup.py (path:/tmp/pip-build-rkyFnu/splunk-sdk/setup.py) egg_info for package splunk-sdk
    
Downloading/unpacking docker-pycreds>=0.2.1 (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
  Downloading docker_pycreds-0.2.1-py2.py3-none-any.whl
Requirement already satisfied (use --upgrade to upgrade): six>=1.4.0 in /usr/lib/python2.7/dist-packages (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
Downloading/unpacking requests>=2.5.2,!=2.11.0 (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
Downloading/unpacking backports.ssl-match-hostname>=3.5 (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
  Downloading backports.ssl_match_hostname-3.5.0.1.tar.gz
  Running setup.py (path:/tmp/pip-build-rkyFnu/backports.ssl-match-hostname/setup.py) egg_info for package backports.ssl-match-hostname
    
Downloading/unpacking ipaddress>=1.0.16 (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
  Downloading ipaddress-1.0.17-py2-none-any.whl
Downloading/unpacking websocket-client>=0.32.0 (from docker-py->-r contrib/loggers/splunk/requirements.txt (line 2))
  Running setup.py (path:/tmp/pip-build-rkyFnu/websocket-client/setup.py) egg_info for package websocket-client
    
Installing collected packages: docker-py, splunk-sdk, docker-pycreds, requests, backports.ssl-match-hostname, ipaddress, websocket-client
  Running setup.py install for splunk-sdk
    
  Found existing installation: requests 2.4.3
    Not uninstalling requests at /usr/lib/python2.7/dist-packages, owned by OS
  Found existing installation: backports.ssl-match-hostname 3.4.0.2
    Not uninstalling backports.ssl-match-hostname at /usr/lib/python2.7/dist-packages, owned by OS
  Running setup.py install for backports.ssl-match-hostname
    
  Found existing installation: websocket-client 0.18.0
    Not uninstalling websocket-client at /usr/lib/python2.7/dist-packages, owned by OS
  Running setup.py install for websocket-client
    changing mode of build/scripts-2.7/wsdump.py from 644 to 755
    
    changing mode of /usr/local/bin/wsdump.py to 755
Successfully installed docker-py splunk-sdk docker-pycreds requests backports.ssl-match-hostname ipaddress websocket-client
Cleaning up...
Sending build context to Docker daemon 27.65 kB
Step 1/5 : FROM outcoldman/splunk:6.4.1
repository outcoldman/splunk not found: does not exist or no read access
---> Making bundle: .integration-daemon-stop (in bundles/1.14.0-dev/test-log-drivers)
+++++ cat bundles/1.14.0-dev/test-log-drivers/docker.pid
++++ kill 10318
++++ /etc/init.d/apparmor stop
Clearing AppArmor profiles cache:.
All profile caches have been cleared, but no profiles have been unloaded.
Unloading profiles will leave already running processes permanently
unconfined, which can lead to unexpected situations.

To set a process to complain mode, use the command line tool
'aa-complain'. To really tear down all profiles, run the init script
with the 'teardown' option."
root@39532f490be8:/go/src/github.com/docker/docker# 
```

**Comment by nwang92** [comment]
@mlaventure Can you try pulling this branch again? The "outcoldman/splunk" reference was changed to "splunk/splunk". You can verify in contrib/loggers/splunk/Dockerfile.

**Comment by mlaventure** [comment]
It looks like the rebase doesn't make the proper merge. Do you mind rebasing yourself first?

**Comment by LK4D4** [comment]
@nwang92 @outcoldman Ok, I certainly do not want to have and run these tests in Docker. I'd prefer splunk to run them on their infrastructure separately(you probably doing it anyway). So I think this should be in separate repo.
ping @thaJeztah @mlaventure

**Comment by LK4D4** [comment]
Okay, no feedback, so I'm closing this :)
Thanks!

**Pull request closed without merging**